packages/@aws-cdk/aws-cloudwatch/lib/metric.ts

```diff
@@ -234,7 +234,21 @@ export class Metric implements IMetric {
    * @param props The set of properties to change.
    */
   public with(props: MetricOptions): Metric {
-    const ret = new Metric({
+    // Short-circuit creating a new object if there would be no effective change
+    if ((props.label === undefined || props.label === this.label)
+      && (props.color === undefined || props.color === this.color)
+      && (props.statistic === undefined || props.statistic === this.statistic)
+      && (props.unit === undefined || props.unit === this.unit)
+      && (props.account === undefined || props.account === this.account)
+      && (props.region === undefined || props.region === this.region)
+      // For these we're not going to do deep equality, misses some opportunity for optimization
+      // but that's okay.
+      && (props.dimensions === undefined)
+      && (props.period === undefined)) {
+      return this;
+    }
```

Comment:
There will never be a bug, just a missed opportunity for an optimization. Metric classes are immutable, always getting a new one (even if you could have reused the old one) is perfectly valid.

They are never checked for object identity equality either, which wouldn't make sense as I can make equivalent objects and they should behave the same in all cases.

What you're saying is correct, but apart from not knowing how I would write an effective regression test in this way, it's also not *that* important. I just wanted to give the garbage collector (and jsii) an easier time.

I can add a test for `period`, given that that's the one that's most likely to change which I wrote this behavior for.

Reply:
I understand the bit about optimization. 

However, when there’s a new property in the Metric class (say `propertyX`) which was missed as part of this `if` statement, it would return an existing `Metric` object without checking whether `propertyX` were equal as well, would you not?